In [8]:
from pymystem3 import Mystem

In [9]:
m = Mystem()

Installing mystem to C:\Users\Дима/.local/bin\mystem.exe from http://download.cdn.yandex.net/mystem/mystem-3.1-win-64bit.zip


In [ ]:
path_to_file = './module16_files/keywords.txt'

In [10]:
i = 0

with open(path_to_file, encoding='utf-8') as f:
    for line in f:
        line = line.strip().split('\t')
        word, wordCount = line
        lemmas = m.lemmatize(word)
        print(word, ''.join(lemmas))
        if i > 10:
            break
        i += 1

вк вк

одноклассники одноклассник

порно порно

ютуб ютуб

вконтакте вконтакте

одноклассники моя страница одноклассник мой страница

майл майл

авито авито

переводчик переводчик

яндекс яндекс

сбербанк онлайн сбербанк онлайн

mail mail



In [14]:
barm_lemm = Mystem()

In [21]:
path_to_file = './module16_files/Бармаглот.txt'

i = 0

with open(path_to_file, encoding='utf-8') as f:
    for line in f:
        line = line.strip()
        lemmas = barm_lemm.lemmatize(line)
        print(''.join(lemmas))
        if i > 3:
            break
        i += 1

варкаться. хливкий шорька

пыряться по нав,

и хрюкотать зелюк,

как мюмзик в мов.




# Упражнение
(1 балл из 1)
Часто вам будут попадаться слова, которые не входят в список словарных. Например, жаргонные или принятые в отрасли слова. Для приведения этих слов к единому виду можно использовать библиотеку Pymystem. У Льюиса Кэррола есть стихотворение "Бармаглот", в котором много несуществующих слов. Проверьте справится ли с лемматизацией строчки из этого стихотворения библиотека Pymystem?

Получите лемму строчки 'и хрюкотали зелюки как мюмзики в мове'. При выводе результата используйте конструкцию print( ''.join( ... ) )

# Библиотека NLTK
Для более быстрой обработки текста рассмотрим библиотеку NLTK (Natural Language Toolkit), которая по умолчанию уже установлена в Anaconda. Библиотека умеет работать со многими языками, а также имеет встроенный список стоп слов, который мы используем в ближайшей задаче. Работает быстрее Pymystem и проводит процедуру стемминга, т. е. процесс нахождения основы слова.

Давайте выведем на экран аналогичные пары слов из нашего файла, но с использованием NLTK:

In [25]:
from nltk.stem import SnowballStemmer

In [26]:
path_to_file = './module16_files/keywords.txt'

In [29]:
sbs = SnowballStemmer('russian')
i = 0

with open(path_to_file, 'r', encoding = 'utf-8' ) as f:
    for line in f:
        line = line.strip().split('\t')          
        stem = sbs.stem(line[0])
        print(line[0], ''.join(stem)) 
        if i > 10:
            break       
        i += 1

вк вк
одноклассники одноклассник
порно порн
ютуб ютуб
вконтакте вконтакт
одноклассники моя страница одноклассники моя страниц
майл майл
авито авит
переводчик переводчик
яндекс яндекс
сбербанк онлайн сбербанк онлайн
mail ma


# Библиотека Pymorphy
Наконец, рассмотрим самый "продвинутый" вариант разбора слов - морфологический анализатор pymorphy. Процесс установки аналогичен:

In [30]:
import pymorphy2

morph = pymorphy2.MorphAnalyzer()

# Рассмотрим принцип ее работы на примере морфологического анализа слова 'стали':

morph.parse('стали')

[Parse(word='стали', tag=OpencorporaTag('VERB,perf,intr plur,past,indc'), normal_form='стать', score=0.984662, methods_stack=((<DictionaryAnalyzer>, 'стали', 904, 4),)),
 Parse(word='стали', tag=OpencorporaTag('NOUN,inan,femn sing,gent'), normal_form='сталь', score=0.003067, methods_stack=((<DictionaryAnalyzer>, 'стали', 13, 1),)),
 Parse(word='стали', tag=OpencorporaTag('NOUN,inan,femn sing,datv'), normal_form='сталь', score=0.003067, methods_stack=((<DictionaryAnalyzer>, 'стали', 13, 2),)),
 Parse(word='стали', tag=OpencorporaTag('NOUN,inan,femn sing,loct'), normal_form='сталь', score=0.003067, methods_stack=((<DictionaryAnalyzer>, 'стали', 13, 5),)),
 Parse(word='стали', tag=OpencorporaTag('NOUN,inan,femn plur,nomn'), normal_form='сталь', score=0.003067, methods_stack=((<DictionaryAnalyzer>, 'стали', 13, 6),)),
 Parse(word='стали', tag=OpencorporaTag('NOUN,inan,femn plur,accs'), normal_form='сталь', score=0.003067, methods_stack=((<DictionaryAnalyzer>, 'стали', 13, 9),))]

Библиотека производит анализ слова и выдает его возможные атрибуты (обратите внимание, что на выходе получился лист из нескольких элементов Parse). Например, первый элемент с параметром score=0.984662 указывается на то, что слово 'стали' скорее всего является глаголом (от слова 'стать'). Тэги в элементе tag указывают на так называемые граммемы, характеризующие данное слово - множественное или единственное число, настоящее или прошедшее время итд. Остальные элементы результата указывают, что слово 'стали' может быть существительным (от слова 'сталь').

Применим аналогичный разбор к первым 5 словам нашего файла:

In [32]:
i = 0

with open(path_to_file, 'r', encoding = 'utf-8' ) as f:
    for line in f:
        line = line.strip().split('\t')
        word = line[0]
        morph_analyze = morph.parse(word)
        print(word, morph_analyze[0].normal_form) 
        if i > 5:
            break       
        i += 1

вк вк
одноклассники одноклассник
порно порно
ютуб ютуба
вконтакте вконтакте
одноклассники моя страница одноклассники моя страница
майл майла


# Упражнение (без проверки)
Дан текст:

In [33]:
words = 'Крымский отель Mriya Resort & Spa признали лучшим в мире курортным комплексом для отдыха по версии престижной международной премии World Travel Awards'

Необходимо вывести на экран существительные в этом тексте.

Подсказка. После получения разбора слова (можно сделать по аналогии с кодом в занятии в переменной morph_analyze) проверить результат на "признак" существительного можно с помощью кода:

In [35]:
for word in words.split():
    morph_analyze = morph.parse(word)
    if 'NOUN' in morph_analyze[0].tag:
        print(word)

отель
мире
комплексом
отдыха
версии
премии
